src/ast/mod.rs

```diff
@@ -779,13 +1037,24 @@ impl Ty {
             TyKind::Ident(_) | TyKind::Dead(_) => false,
         }
     }
+}
 
-    pub fn subst_ident_kinded(&self, ident_kinded: &IdentKinded, with: &ArgKinded) -> Self {
+impl SubstKindedIdents for Ty {
+    fn subst_ident_kinded(&self, ident_kinded: &IdentKinded, with: &ArgKinded) -> Self {
         match &self.ty {
             // TODO mutate and do not create a new type (also this drops the span).
-            TyKind::Data(dty) => Ty::new(TyKind::Data(dty.subst_ident_kinded(ident_kinded, with))),
-            TyKind::Fn(gen_params, params, exec, ret) => Ty::new(TyKind::Fn(
-                gen_params.clone(),
+            TyKind::Data(dty) =>
+                match &dty.dty {
+                    DataTyKind::Ident(ident)
+                        if ident_kinded.kind == Kind::Ty && with.kind() == Kind::Ty && ident.name == ident_kinded.ident.name => 
+                        if let ArgKinded::Ty(ty) = with {
+                            ty.clone()
+                        } else {
+                            panic!("This cannot happen")
+                        },
+                    _ => Ty::new(TyKind::Data(dty.subst_ident_kinded(ident_kinded, with))),
+                },
```

Comment:
Das sieht nicht nach einer so schönen Lösung aus.
Das Problem ist, das ich manchmal Type-Identifier und Data-Type-Identifier habe und ich eigentlich alle Identifier substituiren will.
Beispielsweise sind generischen Parameter Type-Identifier, während die Datentypen in structs beispielsweise DataType-Identifier sind. Und das macht noch ein paar Probleme.
Aber vielleicht kann ich das auch noch irgendwie anders umgehen.

Reply:
Wieso müssen generische Parameter Type Identifier sein?
Ich denke es reicht immer anzunehmen, dass es sich um Data Type Identifier handelt. Ich denke generell gilt, dass Traits (zumindest erstmal) nur für Data Types existieren sollten.
Type Identifier sind nur nötig wenn Funktionen azeptiert werden müssen. Dafür haben wir erstmal kein Trait oder dergleichen.